pkg/openstack/loadbalancer_test.go

```diff
@@ -534,3 +535,292 @@ func Test_getListenerProtocol(t *testing.T) {
+								Address: "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
+							},
+						},
+					},
+				},
+				preferredIPFamily: corev1.IPv4Protocol,
+			},
+			expect:      "192.168.1.1",
+			expectedErr: nil,
+		},
+		{
+			name: "single valid IPv4 address without preferred valid specification",
+			testArgs: testArgs{
+				node: &corev1.Node{
+					Status: corev1.NodeStatus{
+						Addresses: []corev1.NodeAddress{
+							{
+								Type:    corev1.NodeInternalIP,
+								Address: "192.168.1.1",
+							},
+						},
+					},
+				},
+			},
+			expect:      "192.168.1.1",
+			expectedErr: nil,
+		},
+		{
+			name: "single valid IPv6 address without preferred valid specification",
+			testArgs: testArgs{
+				node: &corev1.Node{
+					Status: corev1.NodeStatus{
+						Addresses: []corev1.NodeAddress{
+							{
+								Type:    corev1.NodeInternalIP,
+								Address: "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
+							},
+						},
+					},
+				},
+			},
+			expect:      "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
+			expectedErr: nil,
+		},
+		{
+			name: "multiple valid IPv6 address without preferred valid specification",
+			testArgs: testArgs{
+				node: &corev1.Node{
+					Status: corev1.NodeStatus{
+						Addresses: []corev1.NodeAddress{
+							{
+								Type:    corev1.NodeInternalIP,
+								Address: "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
+							},
+							{
+								Type:    corev1.NodeInternalIP,
+								Address: "192.168.0.1",
+							},
+							{
+								Type:    corev1.NodeInternalIP,
+								Address: "2001:0db8:85a3:1111:2222:8a2e:6869:7334",
+							},
+						},
+					},
+				},
+			},
+			expect:      "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
+			expectedErr: nil,
+		},
+		{
+			name: "invalid IPv4 address specification",
+			testArgs: testArgs{
+				node: &corev1.Node{
+					Status: corev1.NodeStatus{
+						Addresses: []corev1.NodeAddress{
+							{
+								Type:    corev1.NodeInternalIP,
+								Address: "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
+							},
+						},
+					},
+				},
+				preferredIPFamily: corev1.IPv4Protocol,
+			},
+			expect:      "",
+			expectedErr: cpoerrors.ErrNoAddressFound,
+		},
+		{
+			name: "invalid IPv6 address specification",
+			testArgs: testArgs{
+				node: &corev1.Node{
+					Status: corev1.NodeStatus{
+						Addresses: []corev1.NodeAddress{
+							{
+								Type:    corev1.NodeInternalIP,
+								Address: "192.168.1.1",
+							},
+						},
+					},
+				},
+				preferredIPFamily: corev1.IPv6Protocol,
+			},
+			expect:      "",
+			expectedErr: cpoerrors.ErrNoAddressFound,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := nodeAddressForLB(test.testArgs.node, test.testArgs.preferredIPFamily)
+			if err != test.expectedErr {
+				t.Errorf("nodeAddressForLB() error = %v, wantErr %v", err, test.expectedErr)
+				return
+			}
+			if got != test.expect {
+				t.Errorf("nodeAddressForLB() = %v, want %v", got, test.expect)
+			}
```

Comment:
Please make sure to use `assert.Equal()` and `assert.*Error*()` methods instead of manual comparisons.